python/lsst/cp/pipe/utils.py

```diff
@@ -710,6 +711,415 @@ def arrangeFlatsByExpId(exposureList, exposureIdList):
+    ----------
+    diff : `numpy.array`
+        Image where to calculate the covariances (e.g., the difference
+        image of two flats).
+
+    w : `numpy.array`
+        Weight image (mask): it should consist of 1's (good pixel) and
+        0's (bad pixels).
+
+    fftShape : `tuple`
+        2d-tuple with the shape of the FFT
+
+    maxRangeCov : `int`
+        Maximum range for the covariances.
+    """
+
+    def __init__(self, diff, w, fftShape, maxRangeCov):
+        # check that the zero padding implied by "fft_shape"
+        # is large enough for the required correlation range
+        assert(fftShape[0] > diff.shape[0]+maxRangeCov+1)
+        assert(fftShape[1] > diff.shape[1]+maxRangeCov+1)
+        # for some reason related to numpy.fft.rfftn,
+        # the second dimension should be even, so
+        if fftShape[1]%2 == 1:
+            fftShape = (fftShape[0], fftShape[1]+1)
+        tIm = np.fft.rfft2(diff*w, fftShape)
+        tMask = np.fft.rfft2(w, fftShape)
+        # sum of  "squares"
+        self.pCov = np.fft.irfft2(tIm*tIm.conjugate())
+        # sum of values
+        self.pMean = np.fft.irfft2(tIm*tMask.conjugate())
+        # number of w!=0 pixels.
+        self.pCount = np.fft.irfft2(tMask*tMask.conjugate())
+
+    def cov(self, dx, dy):
+        """Covariance for dx,dy averaged with dx,-dy if both non zero.
+
+        Implements appendix of Astier+19.
+
+        Parameters
+        ----------
+        dx : `int`
+           Lag in x
+
+        dy : `int
+           Lag in y
+
+        Returns
+        -------
+        0.5*(cov1+cov2) : `float`
+            Covariance at (dx, dy) lag
+
+        npix1+npix2 : `int`
+            Number of pixels used in covariance calculation.
+        """
+        # compensate rounding errors
+        nPix1 = int(round(self.pCount[dy, dx]))
+        cov1 = self.pCov[dy, dx]/nPix1-self.pMean[dy, dx]*self.pMean[-dy, -dx]/(nPix1*nPix1)
+        if (dx == 0 or dy == 0):
+            return cov1, nPix1
+        nPix2 = int(round(self.pCount[-dy, dx]))
+        cov2 = self.pCov[-dy, dx]/nPix2-self.pMean[-dy, dx]*self.pMean[dy, -dx]/(nPix2*nPix2)
+        return 0.5*(cov1+cov2), nPix1+nPix2
+
+    def reportCovFastFourierTransform(self, maxRange):
+        """Produce a list of tuples with covariances.
+
+        Implements appendix of Astier+19.
+
+        Parameters
+        ----------
+        maxRange : `int`
+            Maximum range of covariances.
+
+        Returns
+        -------
+        tupleVec : `list`
+            List with covariance tuples.
+        """
+        tupleVec = []
+        # (dy,dx) = (0,0) has to be first
+        for dy in range(maxRange+1):
+            for dx in range(maxRange+1):
+                cov, npix = self.cov(dx, dy)
+                if (dx == 0 and dy == 0):
+                    var = cov
+                tupleVec.append((dx, dy, var, cov, npix))
+        return tupleVec
+
+
+def computeCovDirect(diffImage, weightImage, maxRange):
```

Comment:
Is there a reason to keep both the direct and FFT convolutions?  If they give the same result, we should just implement one.  I see we use the FFT by default, so maybe keeping that one?  However, we tend to use lags of 8, which suggests the direct method should be faster (it's also simpler).

Reply:
I kept the FFT version.